source/interaction/interaction_utilities.cc

```diff
@@ -1135,6 +1129,145 @@ namespace fdl
+                                    const unsigned int stencil_axis)
+  {
+    AssertIndexRange(stencil_axis, 3);
+    const dealii::Point<3> p0 = simplex[0];
+    const dealii::Point<3> p1 = simplex[1];
+    const dealii::Point<3> p2 = simplex[2];
+    Tensor<1, 3>           q;
+    q[stencil_axis] = 1;
+    double x_min, x_max, y_min, y_max;
+    x_min = std::numeric_limits<double>::max();
+    x_max = -std::numeric_limits<double>::max();
+    y_min = std::numeric_limits<double>::max();
+    y_max = -std::numeric_limits<double>::max();
+    unsigned int plane_axis[3], i = 0;
+    for (unsigned int axis = 0; axis < 3; ++axis)
+      {
+        if (q[axis] == 0)
+          {
+            plane_axis[i] = axis;
+            i++;
+          }
+      }
+    for (unsigned int nodes = 0; nodes < simplex.size(); ++nodes)
+      {
+        x_min = std::min(simplex[nodes][plane_axis[0]], x_min);
+        x_max = std::max(simplex[nodes](plane_axis[0]), x_max);
+        y_min = std::min(simplex[nodes](plane_axis[1]), y_min);
+        y_max = std::max(simplex[nodes](plane_axis[1]), y_max);
+      }
+    if (!(stencil_start[plane_axis[0]] >= x_min &&
+          stencil_start[plane_axis[0]] <= x_max &&
+          stencil_start[plane_axis[1]] <= y_max &&
+          stencil_start[plane_axis[1]] >= y_min))
+      {
+        return std::nullopt;
+      }
+    const Tensor<1, 3> e1 = p1 - p0;
+    const Tensor<1, 3> e2 = p2 - p0;
+    const Tensor<1, 3> h  = cross_product_3d(q, e2);
+    double             a  = e1 * h;
+    if (std::abs(a) > std::numeric_limits<double>::epsilon())
+      {
+        double             f = 1.0 / a;
+        const Tensor<1, 3> s = stencil_start - p0;
+        double             u = f * (s * h);
+        if (u >= 0 - std::numeric_limits<double>::epsilon() &&
+            u <= 1.0 + std::numeric_limits<double>::epsilon())
+          {
+            const Tensor<1, 3> cr = cross_product_3d(s, e1);
+            double             v  = f * (q * cr);
+            if (v > 0 - std::numeric_limits<double>::epsilon() &&
+                (u + v) <= 1.0 + std::numeric_limits<double>::epsilon())
+              {
+                double       t           = f * (e2 * cr);
+                const double convex_coef = t / stencil_width;
+
+                return abs(convex_coef) <= 1 ?
+                         std_cxx17::optional<double>(convex_coef) :
+                         std::nullopt;
+              }
+          }
+      }
+    return std::nullopt;
+  }
+
+
+  template <int spacedim>
+  std_cxx17::optional<double>
+  intersect_stencil_with_simplex(
+    const std::array<Point<spacedim + 1>, spacedim + 1> &simplex,
+    const Point<spacedim + 1>                           &stencil_start,
+    const double                                        &stencil_width,
+    const unsigned int                                   stencil_axis)
+  {
+    Assert(stencil_width > 0.0, ExcMessage("stencil width should be positive"));
+#if (NDIM == 3)
```

Comment:
This is OK for now but I'll fix it in a followup PR - there's no need to use the preprocessor for these things.